### PR TITLE
Added option to throttle frame rate and trigger single images

### PR DESCRIPTION
**Pull request by jwendelmuth**
Added ability to throttle framerate using a parameter /rviz_camera_framerate 
(Absolute name as RVIZ is started anonymously and other nodes need a fixed name).
If framerate is set to 0, no images are published.

It is also possible to trigger a single image by sending
 a std_srvs/Trigger on /rviz_camera_trigger.

### CODE REVIEW

File: include/rviz_camera_stream/camera_display.h [hunk 1]
```diff
@@ -38,9 +38,10 @@
 #include <OgreSharedPtr.h>
 #include <OgreTexture.h>
 
-# include <sensor_msgs/CameraInfo.h>
+#include <sensor_msgs/CameraInfo.h>
```
Comment (lucasw) [inline include/rviz_camera_stream/camera_display.h]: Put format changes in a different pull request

File: include/rviz_camera_stream/camera_display.h [hunk 2]
```diff
@@ -107,6 +108,17 @@ private Q_SLOTS:
   void subscribe();
   void unsubscribe();
 
+  ros::ServiceServer trigger_service_;
+  bool triggerCallback(std_srvs::TriggerRequest& req,
+                       std_srvs::TriggerResponse& res)
+  {
+      trigger_activated_ = true;
```
Comment (lucasw) [inline include/rviz_camera_stream/camera_display.h]: put the code in the cpp file

File: src/camera_display.cpp [hunk 3]
```diff
@@ -254,8 +258,20 @@ void CameraPub::preRenderTargetUpdate(const Ogre::RenderTargetEvent& evt)
 
 void CameraPub::postRenderTargetUpdate(const Ogre::RenderTargetEvent& evt)
 {
-  // TODO(lucasw) allow this to be throttled down
   // Publish the rendered window video stream
+  float framerate = -1;
+  ros::param::get("/rviz_camera_framerate", framerate); /// Absolute name as RVIZ is started anonymously and other nodes need a fixed name
```
Comment (lucasw) [inline src/camera_display.cpp]: The parameter server is meant to be used on initialization or otherwise infrequently, so only get the parameter in onInitialize()- if the framerate needs to be changed live do it with a service callback or even a topic (an rviz plugin gui control made using a float property would be nice too).

File: src/camera_display.cpp [hunk 4]
```diff
@@ -254,8 +258,20 @@ void CameraPub::preRenderTargetUpdate(const Ogre::RenderTargetEvent& evt)
...
+  ros::param::set("/rviz_camera_framerate", framerate); // make sure that parameter is shown in rosparam list
+
+  static ros::Time last_image_time = ros::Time::now();
```
Comment (lucasw) [inline src/camera_display.cpp]: Make this a class member variable instead of static.

File: src/camera_display.cpp [hunk 5]
```diff
@@ -254,8 +258,20 @@ void CameraPub::preRenderTargetUpdate(const Ogre::RenderTargetEvent& evt)
...
+  bool time_is_up = (framerate > 0) && (ros::Time::now() - last_image_time).toSec() > 1.0 / framerate;
```
Comment (lucasw) [inline src/camera_display.cpp]: Instead of calling ros::Time::now() over and over, use last_image_time

**Comment by lucasw** [comment]
It looks like this can only have framerates that can be multiplied by an integer to equal ~30 (the base rviz framerate)- like 1,2,3, 5,6, 10, and 15 but nothing in-between.

That might be desirable in cases where a fixed interval between frames is desired, another approach could be to update 'last_publication_time_' to last_publication_time_ + 1.0 / frame_rate instead of the current time- the output frame rate will be uneven but average out to closer to what the user wants.  (The initial value of last_publication_time_ would need to be set to the initialization time rather than zero also)  (Also #7 would go well with that change.)  Maybe there could be a checkbox to switch behaviors.

That isn't necessary to add for this merge request, I'm just recording the idea here.  I'll probably make a few edits and then merge this either today or Monday.

**Comment by lucasw** [comment]
I made some additional formatting changes and squashed into one commit 7839c3a , thanks!

As far as rviz being launch anonymous, it is a better practice to launch it with a name and then multiple rviz instance can exist without having to share the same trigger service call, so I'll likely change that to a non-absolte service later. (does it even work to have two nodes with the same service?  Do they both get called?).

**Comment by lucasw** [comment]
Not sure if there is a more proper way to edit and merge and close the way I did and make the pull request look like it was merged.

Also if you make additional changes you'll have to clobber your current master with my master and lose your old history, it's better to have one named feature branch per pull request and have your master track upstream.

**Comment by NikolasE** [comment]
" Do they both get called?)."

That's what I also thought about this weekend. I think we should add a name so that we can use several camera streams (could be very useful to create a video with a bird's eye view and one view from the robot)

**Comment by lucasw** [comment]
It certainly doesn't work to have the same node with two instances trying to advertise the same service: `[ERROR] [/rviz] [/tmp/binarydeb/ros-kinetic-roscpp-1.12.2/src/libros/service_manager.cpp]:[147] [Tried to advertise a service that is already advertised in this node [/rviz_camera_trigger]]`

**Comment by NikolasE** [comment]
I created a new PR to deal with this issue
